### ISSUE TITLE
Offcanvas Pro Menu doesn't close when using hash links

### ISSUE THREAD

**Issue by sunnyt7** [soflyy/oxygen-bugs-and-features#1022]
This is in 3.3 alpha, but existed previously as well.

1. Create a WordPress menu with a link called "About" and the URL "#about"
2. Create a Pro Menu with the offcanvas option enabled, and select the menu you created
3. Preview your site on the front-end, and click on your About link. The page will scroll to the appropriate section, but the offcanvas pro menu does not close as it should. If you try this on mobile, it will feel like clicking the button didn't actually do anything.

Someone in the group has provided a script that fixes this issue:

`jQuery(document).ready(function($) {
  $("body").on("click", ".oxy-pro-menu-list > li > a", function(e) {
    let currentPathname = window.location.href.split("://").pop();
    let urlClickedOn = this.href.split("://").pop();

    if (currentPathname.split("#").shift() == urlClickedOn.split("#").shift()) {
      $(this).parents('.oxy-pro-menu').removeClass('oxy-pro-menu-open');
      $(this).parents('.oxy-pro-menu-container').removeClass('oxy-pro-menu-open-container');
      $('.oxy-nav-menu-prevent-overflow').removeClass('oxy-nav-menu-prevent-overflow');

      if ($(this).parents('.oxy-pro-menu-container').hasClass('oxy-pro-menu-off-canvas-container')) {
        oxygen_offcanvas_menu_run($(this).parents('.oxy-pro-menu'));
      }
    }
  });
});`

**Comment by adrien-robert** [comment]
+1 for a fix in Pro Menu itself (I'm the one who provided the temporary fix)

**Comment by runechronos** [comment]
Hy, @adrien-robert your fix does not seem to work with sub-menu items, when I click on the icon to open my sub menu under a menu item, it closes the menu instead of just showing the sub menu. 
Because it's not differenciating the click on the "a" tag and the click on the dropdown icon inside the a tag

So your condition : 

**` $("body").on("click", ".oxy-pro-menu-list li > a", function(e)`**

Might not be enough, you should put `**li a** not **li > a**` to also target submenus, but I can't figure how to differenciate the click on the **"a"** tag and on the icon as I said, if you have any idea, let me know ! 

But it is working with just a menu with items without subitems.

**Comment by adrien-robert** [comment]
> Hy, @adrien-robert your fix does not seem to work with sub-menu items, when I click on the icon to open my sub menu under a menu item, it closes the menu instead of just showing the sub menu.
> Because it's not differenciating the click on the "a" tag and the click on the dropdown icon inside the a tag
> 
> So your condition :
> 
> **` $("body").on("click", ".oxy-pro-menu-list li > a", function(e)`**
> 
> Might not be enough, you should put `**li a** not **li > a**` to also target submenus, but I can't figure how to differenciate the click on the **"a"** tag and on the icon as I said, if you have any idea, let me know !
> 
> But it is working with just a menu with items without subitems.

Yeah I just coded this for someone in Oxygen FB group so it's not for all cases ahah ^^'
You can be free to modify!

**Comment by runechronos** [comment]
@adrien-robert yes :), I'm trying but I don't get how to differenciate clicks on `<a> tag and on <svg> inside the same <a> tag, as you are always clicking on the <a> anyway.` :s

**Comment by runechronos** [comment]
I found a way to tweak you previous script to make it work for any menu with submenus. I post it here if someone encounter the same problem:

`jQuery(document).ready(function($) {
  $("body").on("click", ".oxy-pro-menu-list > li a", function(e) {
    let currentPathname = window.location.href.split("://").pop();
    let urlClickedOn = this.href.split("://").pop();

    
    if ($(this).children('.oxy-pro-menu-dropdown-icon-click-area').children('.oxy-pro-menu-dropdown-icon').length == 0) {
      if (currentPathname.split("#").shift() == urlClickedOn.split("#").shift()) {



          $(this).parents('.oxy-pro-menu').removeClass('oxy-pro-menu-open');
          $(this).parents('.oxy-pro-menu-container').removeClass('oxy-pro-menu-open-container');
          $('.oxy-nav-menu-prevent-overflow').removeClass('oxy-nav-menu-prevent-overflow');

          if ($(this).parents('.oxy-pro-menu-container').hasClass('oxy-pro-menu-off-canvas-container')) {
            oxygen_offcanvas_menu_run($(this).parents('.oxy-pro-menu'));
          }
      }
    }
  });
});`

**Comment by katsar0v** [comment]
Is there a better, maybe official solution, ping @KittenCodes ?

**Comment by KittenCodes** [comment]
@katsar0v This should do the trick - please try it in the JavaScript tab of a Code Block element:

```js
jQuery(document).ready(function() { 
  jQuery('a[href^="#"]').click(function() { 
    jQuery('.oxy-pro-menu-mobile-close-icon').trigger('click');
  }); 
}); 
```

**Comment by leem2209** [comment]
I've just noticed this issue. Is it still a bug? Can I ask how people are adding the codes above? I've added a code block (into my header template) and used the jQuery code above but it doesn't seem to make any difference.

**Comment by leem2209** [comment]
Did anyone find a way to fix this? The code suggested by @KittenCodes doesn't seem to be working for me.

**Comment by runechronos** [comment]
@KittenCodes your solution is not working properly, it works the 1st time u click on the anchor menu item, but when u do it a second time:

It triggers an infinite loop in js console: _too much recursion_

Do you have an idea on how to solve this ?

**Comment by runechronos** [comment]
No news ? Is no one able to answer me on the team ? :s

**Comment by leem2209** [comment]
Support looked at this for me and eventually we got this to work.

Insert a code block at the very top of your header/footer template file. Delete the PHP code and in the JS code, add this:

`jQuery(document).ready(function() {
    jQuery('.oxy-pro-menu .menu-item a[href*="#"]').on('click', function (e) {
        jQuery('.oxy-pro-menu-mobile-close-icon').trigger('click');
    });
});`

**Comment by runechronos** [comment]
@leem2209 thanks for you answer, the tweak is indeed workign but is still problematic, because if you have a menu with some submenus links, and you don't want the parend of the subfolder to be clickable u usually set it's url to # in wordpress.
So because of this each time you click on the parent to show the submenu items, it removes the menu from screen as there is # in its URL as in the anchor links :s. Trying to add some conditions to check the length after the # maybe.

**Comment by AlkoKod** [comment]
These codes need to be little bit modified. I found the last entry from @runechronos working well, but how to make it work if I use full URL address in menu link? Like https://www.untitled.com/#contact ? Is it possible?

**Comment by yankiara** [comment]
@KittenCodes: Here's an idea of quick fix...

It might still require some hashlink detection (length>1) but it works for me.

In file /component-framework/components/classes/menu-pro.class.php, just insert this code at line 1766:
```
if (jQuery(this).attr("href").indexOf("#")===0 && !(jQuery(this).closest(".oxy-pro-menu-container").data("entire-parent-toggles-dropdown") && jQuery(this).parent().hasClass('menu-item-has-children'))) {
	var menu = jQuery(this).parents('.oxy-pro-menu');
	oxygen_offcanvas_menu_run(menu);
	oxygen_pro_menu_unset_static_width(menu);
}
```

This fix also takes "Entire parent toggles dropdown" option into account (see my fix for #2491).

**Comment by yankiara** [comment]
@KittenCodes: I actually simplified the whole menu link click code and this is the result:

```
			// Menu link click (standard and offcanvas merged)
			var selector = '.oxy-pro-menu-open .menu-item a, .oxy-pro-menu-off-canvas .menu-item a';
			jQuery('body').on('click', selector, function(event){

				if (jQuery(event.target).closest('.oxy-pro-menu-dropdown-icon-click-area').length > 0) {
					// toggle icon clicked
					return;
				}
				else if ((jQuery(this).attr("href") === "#" || jQuery(this).closest(".oxy-pro-menu-container").data("entire-parent-toggles-dropdown")) && jQuery(this).parent().hasClass('menu-item-has-children')) {
					// empty href don't lead anywhere, treat it as toggle trigger
					oxygen_pro_menu_toggle_dropdown(event.target);
					// keep anchor links behavior as is, and prevent regular links from page reload
					if (jQuery(this).attr("href").indexOf("#")!==0) {
						return false;
					}
				}

				// hide the menu and follow the anchor
				if (jQuery(this).attr("href").indexOf("#")===0 && !(jQuery(this).closest(".oxy-pro-menu-container").data("entire-parent-toggles-dropdown") && jQuery(this).parent().hasClass('menu-item-has-children'))) {
					jQuery(".oxy-pro-menu-mobile-close-icon").trigger("click"); 
				}

			});
```

Just replace lines 1723 to 1766 with this code :-)

It fixes #2491 as well.

**Comment by KittenCodes** [comment]
Thanks @yankiara. I've opened an internal issue that links to the above code.

**Comment by AlkoKod** [comment]
Hello @yankiara and @KittenCodes. I tried to use suggested codes, but I can't make it work properly. 

I have my anchor links added with full domain name, because I am not on one page website. When user go to the homepage, open mobile menu and click on link - for example: home.com/#about - then nothing happen and menu stays opened. 

Can I fix this?
